.github/workflows/tests.yml

```diff
@@ -0,0 +1,87 @@
+name: Test
+
+on:
+  push:
+    branches:
+      - "**"
+  pull_request:
+    branches:
+      - main
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    if: >
+      (github.event_name == 'push' && github.ref != 'refs/heads/main') ||
+      (github.event_name == 'pull_request' && github.base_ref == 'main')
+    container:
+      image: python:3.12-slim
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Install dependencies
+        shell: bash
+        run: |
+          apt-get update && apt-get install -y shellcheck make
+          python -m venv venv
```

Comment:
Somewhat annoyingly for the ci/cd case, the make tests require everything to run in a venv, so we have to set that up, and repeat it in every step. We could put all the steps together but then the debugging is more of a pain, so I guess the redundant code is better.

Reply:
You could also run:
```
make venv
make install
```
That should do the same thing. But what you have is fine